metadata-ingestion/src/datahub/utilities/logging_manager.py

```diff
@@ -0,0 +1,194 @@
+                return record.levelno >= logging.DEBUG
+            else:
+                return record.levelno >= logging.INFO
+        else:
+            if self.debug:
+                return record.levelno >= logging.WARNING
+            else:
+                return record.levelno >= logging.INFO
+
+
+class _LogBuffer:
+    def __init__(self, maxlen: Optional[int] = None) -> None:
+        self._buffer: Deque[str] = collections.deque(maxlen=maxlen)
+
+    def write(self, line: str) -> None:
+        self._buffer.append(line)
+
+    def clear(self) -> None:
+        self._buffer.clear()
+
+    def format_lines(self) -> str:
+        text = "\n".join(self._buffer)
+        if len(self._buffer) == 0:
+            text = "[no logs]"
+        elif len(self._buffer) == self._buffer.maxlen:
+            text = f"[earlier logs truncated...]\n{text}"
+        return text
+
+
+class _BufferLogHandler(logging.Handler):
+    def __init__(self, storage: _LogBuffer) -> None:
+        super().__init__()
+        self._storage = storage
+
+    def emit(self, record: logging.LogRecord) -> None:
+        self._storage.write(self.format(record))
+
+
+def _remove_all_handlers(logger: logging.Logger) -> None:
+    for handler in logger.handlers[:]:
+        logger.removeHandler(handler)
+        handler.close()
+
+
+_log_buffer = _LogBuffer(maxlen=IN_MEMORY_LOG_BUFFER_SIZE)
+
+
+def get_log_buffer() -> _LogBuffer:
+    return _log_buffer
+
+
+_stream_formatter = _ColorLogFormatter()
+_default_formatter = logging.Formatter(BASE_LOGGING_FORMAT)
+
+
+@contextlib.contextmanager
+def configure_logging(debug: bool, log_file: Optional[str] = None) -> Iterator[None]:
+    _log_buffer.clear()
+
+    with contextlib.ExitStack() as stack:
+        # Create stdout handler.
+        stream_handler = logging.StreamHandler()
+        stream_handler.addFilter(_DatahubLogFilter(debug=debug))
+        stream_handler.setFormatter(_stream_formatter)
+
+        # Create file handler.
+        file_handler: logging.Handler
+        if log_file:
+            file = stack.enter_context(open(log_file, "w"))
+            tee = TeeIO(sys.stdout, file)
+            stack.enter_context(contextlib.redirect_stdout(tee))  # type: ignore
+
+            file_handler = logging.StreamHandler(file)
+            file_handler.addFilter(_DatahubLogFilter(debug=True))
+            file_handler.setFormatter(_default_formatter)
+        else:
+            file_handler = logging.NullHandler()
+
+        # Create the in-memory buffer handler.
+        buffer_handler = _BufferLogHandler(_log_buffer)
+        buffer_handler.addFilter(_DatahubLogFilter(debug=debug))
+        buffer_handler.setFormatter(_default_formatter)
+
+        handlers = [
+            stream_handler,
+            file_handler,
+            buffer_handler,
+        ]
+
+        # Configure the loggers.
+        root_logger = logging.getLogger()
+        _remove_all_handlers(root_logger)
+        root_logger.setLevel(logging.INFO)
```

Comment:
why do we need this "normalization" step?

Reply:
the default is warning globally, and can also be messed with by other libraries